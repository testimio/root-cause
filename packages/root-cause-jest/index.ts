export * from './lib/helpers';
export * from './lib/jasmine2JestRelated';
