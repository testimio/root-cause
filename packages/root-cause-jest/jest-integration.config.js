'use strict';

/**
 * The reason we need separate jest config, and to run jest separately,
 * Is that this test when ran with the rest of the tests gets everything crazy on circle ci
 * locally it was passing
 */
module.exports = {
  preset: 'ts-jest',
  testMatch: ['<rootDir>/lib/integration.test.ts'],
};
