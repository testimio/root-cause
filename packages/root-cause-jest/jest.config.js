'use strict';

module.exports = {
  preset: 'ts-jest',
  // This is here to avoid running built tests on dist/ dir
  roots: ['<rootDir>/lib/'],
  testPathIgnorePatterns: ['/node_modules/', '.js$', '<rootDir>/lib/integration.test.ts'],
};
