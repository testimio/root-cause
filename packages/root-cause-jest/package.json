{
  "name": "@testim/root-cause-jest",
  "version": "0.1.22",
  "author": "Testim.io <dev@testim.io>",
  "engines": {
    "node": ">= 10.0.0"
  },
  "scripts": {
    "test": "jest -c jest.config.js",
    "test:integration": "jest -c jest-integration.config.js",
    "build": "tsc && npx downlevel-dts ./dist/ ./dist/ts3.4",
    "prepublishOnly": "./scripts/prepublishOnly.sh",
    "root-cause": "ts-node ../root-cause-core/lib/cli",
    "rc": "ts-node ../root-cause-core/lib/cli"
  },
  "main": "./index",
  "license": "AGPL-3.0-only",
  "typesVersions": {
    "<3.8": {
      "*": [
        "ts3.4/*"
      ]
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/testimio/root-cause.git"
  },
  "bugs": {
    "url": "https://github.com/testimio/root-cause/issues"
  },
  "homepage": "https://help.testim.io/docs/testim-root-cause-overview",
  "peerDependencies": {
    "@jest/reporters": "^26.4.1 || ^25.5.1"
  },
  "devDependencies": {
    "@jest/fake-timers": "^26.5.2",
    "@jest/reporters": "^26.5.0",
    "@jest/types": "^26.5.2",
    "@testim/root-cause-types": "0.1.22",
    "@types/debug": "^4.1.5",
    "@types/fs-extra": "^9.0.1",
    "@types/jest": "^26.0.14",
    "@types/uuid": "^8.3.0",
    "jest-mock": "^26.5.2",
    "ts-node": "^9.0.0",
    "typescript": "^3.9.7"
  },
  "dependencies": {
    "@testim/root-cause-core": "0.1.22",
    "chalk": "^4.1.0",
    "debug": "^4.1.1",
    "fs-extra": "^9.0.1"
  },
  "publishConfig": {
    "directory": "dist",
    "access": "public"
  }
}
