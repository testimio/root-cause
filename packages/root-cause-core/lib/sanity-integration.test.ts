import { attach } from './index';
import puppeteer from 'puppeteer';
import fs from 'fs-extra';
import path from 'path';
import assert from 'assert';
import {
  assertNotNullOrUndefined,
  testUniqueIdentifierFromStartParams,
  jsonReduceNoiseReviver,
} from './utils';
import {
  getCleanAllPathsPrettyFormatPlugin,
  getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin,
} from '@testim/internal-self-tests-helpers';
// eslint-disable-next-line import/no-extraneous-dependencies
import { registerJasmineReporterToGlobal, getJasmineCurrentTest } from '@testim/root-cause-jest';
import type { TestResultFile } from '@testim/root-cause-types';

describe('Sanity integration test', () => {
  jest.setTimeout(30_000);
  expect.addSnapshotSerializer(getCleanAllPathsPrettyFormatPlugin(process.cwd()));
  expect.addSnapshotSerializer(getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin());

  let browser: puppeteer.Browser;
  let browserContext: puppeteer.BrowserContext;
  let page: puppeteer.Page;

  beforeAll(async () => {
    registerJasmineReporterToGlobal();

    // https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox
    // Not optimal, but didn't work on circle on circleci/node:12.17-stretch-browsers without it
    // Need to revisit
    browser = await puppeteer.launch({ args: ['--no-sandbox', '--disable-setuid-sandbox'] });
  });

  afterAll(async () => {
    await browser.close();
  });

  beforeEach(async () => {
    browserContext = await browser.createIncognitoBrowserContext();
    page = await browserContext.newPage();
  });

  afterEach(async () => {
    await page.close();
    await browserContext.close();
  });

  it('Sanity integration test 1', async () => {
    const currentTestInfo = getJasmineCurrentTest();

    const startTestParams = {
      runId: 'mock_invocation_id',
      projectRoot: path.resolve(__dirname, 'testsResults'),
      fullName: currentTestInfo.fullName,
      description: currentTestInfo.description,
      fullSuitePath: __filename,
    };

    const mockedDateConstructor: typeof Date = Object.create(Date);

    let nowCallsCounter = 1;
    mockedDateConstructor.now = function mockedNow() {
      return nowCallsCounter++;
    };

    const { page: playedPage, endTest } = await attach(
      {
        page,
        startTestParams,
      },
      mockedDateConstructor
    );

    await playedPage.goto('http://jsbin.testim.io/ner/1');
    await playedPage.click('#test');
    // to await - might be flaky
    page.evaluate(() => {
      setTimeout(() => {
        throw new Error('injected error');
      }, 0);
    });
    await playedPage.click('#test');
    await playedPage.click('#test');
    // to await - might be flaky
    page.evaluate(() => {
      setTimeout(() => {
        /* eslint-disable-next-line no-console */
        console.warn('injected console');
      }, 0);
    });
    await playedPage.click('#test');
    const counterEl = await playedPage.$('#counter');

    assert.notEqual(counterEl, null);
    assertNotNullOrUndefined(counterEl);

    const text = await (await counterEl.getProperty('innerText')).jsonValue();

    assert.equal(text, '4');

    await playedPage.click('#reset');

    const text2 = await (await counterEl.getProperty('innerText')).jsonValue();

    assert.equal(text2, '0');

    const allButtons = await playedPage.$$('button');

    await allButtons[0].click();

    const text3 = await (await counterEl.getProperty('innerText')).jsonValue();

    assert.equal(text3, '1');

    assert.equal(allButtons.length, 2);

    await endTest({
      success: false,
      error: { message: 'mocked error test failed' },
    });

    const expectedResultsPath = path.resolve(
      __dirname,
      'testsResults',
      '.root-cause',
      'runs',
      'mock_invocation_id',
      testUniqueIdentifierFromStartParams(startTestParams)
    );
    const testResults: TestResultFile = JSON.parse(
      await fs.readFile(path.resolve(expectedResultsPath, 'results.json'), 'utf-8'),
      jsonReduceNoiseReviver
    );

    expect(testResults).toMatchSnapshot();

    for (const stepResult of testResults.steps) {
      // Do also images blob comparison? TBH
      if (stepResult.screenshot) {
        assert.equal(
          await fs.pathExists(path.resolve(expectedResultsPath, stepResult.screenshot)),
          true,
          `${stepResult.screenshot} is missing`
        );
      }
    }
  });
});
