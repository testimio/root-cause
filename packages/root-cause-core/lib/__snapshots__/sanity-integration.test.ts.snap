// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity integration test Sanity integration test 1 1`] = `
Object {
  "metadata": Object {
    "branchInfo": "noise_removed:branchInfo",
    "endedTimestamp": 27,
    "fileName": "noise_removed/packages/root-cause-core/lib/sanity-integration.test.ts",
    "hasNetworkLogs": true,
    "systemInfo": Object {
      "automationFramework": "puppeteer",
      "browser": "chromium",
      "browserPlatform": "noise_removed:browserPlatform",
      "browserVersion": "HeadlessChrome/85.0.4182.0",
      "modelName": "noise_removed:modelName",
      "modelVersion": "noise_removed:modelVersion",
      "pageViewport": Object {
        "height": 600,
        "width": 800,
      },
      "userAgent": "noise_removed:userAgent",
    },
    "testEndStatus": Object {
      "error": Object {
        "message": "mocked error test failed",
        "name": "",
        "stack": "",
      },
      "success": false,
    },
    "testFullName": "Sanity integration test Sanity integration test 1",
    "testName": "Sanity integration test 1",
    "timestamp": 2,
  },
  "steps": Array [
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 4,
      "fnName": "goto",
      "index": 1,
      "name": "goto \\"http://jsbin.testim.io/ner/1\\"",
      "screenshot": "1.page-goto.jpg",
      "startTimestamp": 3,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "            startTestParams,",
          "        }, mockedDateConstructor);",
          "",
          "        await playedPage.goto('http://jsbin.testim.io/ner/1');",
          "        await playedPage.click('#test');",
          "        // to await - might be flaky",
          "        page.evaluate(() => {",
        ],
        "column": 9,
        "fromRowNumber": 64,
        "row": 67,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 70,
      },
      "text": "http://jsbin.testim.io/ner/1",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 6,
      "fnName": "click",
      "index": 2,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "2.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 5,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "        }, mockedDateConstructor);",
          "",
          "        await playedPage.goto('http://jsbin.testim.io/ner/1');",
          "        await playedPage.click('#test');",
          "        // to await - might be flaky",
          "        page.evaluate(() => {",
          "            setTimeout(() => {",
        ],
        "column": 9,
        "fromRowNumber": 65,
        "row": 68,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 71,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 9,
      "fnName": "click",
      "index": 3,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "3.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 7,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                throw new Error('injected error');",
          "            }, 0);",
          "        });",
          "        await playedPage.click('#test');",
          "        await playedPage.click('#test');",
          "        // to await - might be flaky",
          "        page.evaluate(() => {",
        ],
        "column": 9,
        "fromRowNumber": 72,
        "row": 75,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 78,
      },
      "unhandledExceptions": Array [
        Object {
          "message": "injected error",
          "stack": "Error: injected error
    at __puppeteer_evaluation_script__:3:23",
          "timestamp": 8,
        },
      ],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 11,
      "fnName": "click",
      "index": 4,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "4.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 10,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "            }, 0);",
          "        });",
          "        await playedPage.click('#test');",
          "        await playedPage.click('#test');",
          "        // to await - might be flaky",
          "        page.evaluate(() => {",
          "            setTimeout(() => {",
        ],
        "column": 9,
        "fromRowNumber": 73,
        "row": 76,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 79,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [
        Object {
          "args": Array [
            "\\"injected console\\"",
          ],
          "column": 24,
          "level": "warning",
          "line": 3,
          "text": "injected console",
          "timestamp": 13,
          "url": "__puppeteer_evaluation_script__",
        },
      ],
      "endTimestamp": 14,
      "fnName": "click",
      "index": 5,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "5.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 12,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "                console.warn('injected console');",
          "            }, 0);",
          "        });",
          "        await playedPage.click('#test');",
          "        const counterEl = await playedPage.$('#counter');",
          "",
          "        assert.notEqual(counterEl, null);",
        ],
        "column": 9,
        "fromRowNumber": 81,
        "row": 84,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 87,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 16,
      "fnName": "getProperty",
      "index": 6,
      "name": "getProperty",
      "screenshot": "6.elementhandle-getProperty.jpg",
      "startTimestamp": 15,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "        assert.notEqual(counterEl, null);",
          "        assertNotNullOrUndefined(counterEl);",
          "",
          "        const text = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "        assert.equal(text, '4');",
          "",
        ],
        "column": 29,
        "fromRowNumber": 87,
        "row": 90,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 93,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 18,
      "fnName": "click",
      "index": 7,
      "name": "click \\"#reset\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "7.page-click.jpg",
      "selector": "#reset",
      "startTimestamp": 17,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "",
          "        assert.equal(text, '4');",
          "",
          "        await playedPage.click('#reset');",
          "",
          "        const text2 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
        ],
        "column": 9,
        "fromRowNumber": 91,
        "row": 94,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 97,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 20,
      "fnName": "getProperty",
      "index": 8,
      "name": "getProperty",
      "screenshot": "8.elementhandle-getProperty.jpg",
      "startTimestamp": 19,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "",
          "        await playedPage.click('#reset');",
          "",
          "        const text2 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "        assert.equal(text2, '0');",
          "",
        ],
        "column": 30,
        "fromRowNumber": 93,
        "row": 96,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 99,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 22,
      "fnName": "click",
      "index": 9,
      "name": "click",
      "screenshot": "9.elementhandle-click.jpg",
      "startTimestamp": 21,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "",
          "        const allButtons = await playedPage.$$('button');",
          "",
          "        await allButtons[0].click();",
          "",
          "        const text3 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
        ],
        "column": 9,
        "fromRowNumber": 99,
        "row": 102,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 105,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 24,
      "fnName": "getProperty",
      "index": 10,
      "name": "getProperty",
      "screenshot": "10.elementhandle-getProperty.jpg",
      "startTimestamp": 23,
      "stepCodeLocation": Object {
        "codeLines": Array [
          "",
          "        await allButtons[0].click();",
          "",
          "        const text3 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "        assert.equal(text3, '1');",
          "",
        ],
        "column": 30,
        "fromRowNumber": 101,
        "row": 104,
        "sourceFileRelativePath": "lib/sanity-integration.test.ts",
        "toRowNumber": 107,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "endTimestamp": 26,
      "fnName": "Assertion Error",
      "index": 11,
      "name": "Assertion Error \\"mocked error test failed\\"",
      "rect": Object {
        "error": "possible noise removed",
      },
      "screenshot": null,
      "startTimestamp": 25,
      "stepError": Object {
        "message": "mocked error test failed",
        "name": "",
        "stack": "",
      },
      "text": "mocked error test failed",
    },
  ],
}
`;
